feat(components): new Modal Page design

jonnybel opened the pull request:
#### Summary

This PR reimplements the stacking Modal Pages according to the new designs.

tdeekens commented on packages/application-components/src/components/modal-page/README.md:
```diff
@@ -14,41 +14,42 @@ Modal pages are controlled components used to render a page using a modal contai
-    ),
-  }}
+  headerActions={
+    <Spacings.Inline alignItems="flex-start">
```
Didnt we agree that splitting actions into two would be good?

jonnybel replied:
Yes, however after further analysing the designs and use cases for this, I found cases where there are 3 different buttons, and where the type of the buttons (primary/secondary, iconButton) also differs.
So, given this need for flexibility which would require a bunch of different parameters, I decided to keep the actions has a render prop.

tdeekens replied:
> actions has a render prop.

It's not a render-prop. That would imply it being an fn receiving arguments. Is that what we need here?

jonnybel replied:
Sorry, you're right, not a render prop. It's a node which its rendered aligned in the right side of the Header section of the Modal.

montezume commented on packages/application-components/src/components/modal-page/internals/modal-page-top-bar.js:
```diff
@@ -0,0 +1,48 @@
+const ModalPageTopBar = props => (
+  <div css={getTopBarStyles(props)}>
+    <Spacings.Inline>
+      <FlatButton
```
from the designs it looks like there can be an array of previous paths right?

jonnybel replied:
At first I thought so, like breadcrumbs, but then I didn't find any other than just the previous path and the current one. I'm going to clarify this.

jonnybel replied:
Actually, on another look I've found sometimes there are more than just one previous paths, so I'm going to refactor this to use an array for such.

jonnybel replied:
After discussing with design, this is only supposed to act like a "Back Button". There are plans to implement a "breadcrumbs" component but not for this phase.

tdeekens commented on packages/application-components/src/components/modal-page/internals/modal-page-top-bar.js:
```diff
@@ -0,0 +1,48 @@
+    <Spacings.Inline>
+      <FlatButton
+        tone="primary"
+        label={props.previousPath || ''}
```
defaultProps instead?

jonnybel replied:
Ah right, and I'm actually setting it as _requiredIf showTopBar = true_

jonnybel replied:
Talked with design and this now defaults to "Go Back" (translated)

tdeekens commented on packages/application-components/src/components/modal-page/modal-page.js:
```diff
@@ -29,136 +28,74 @@ const parentSelector = () =>
-  components: PropTypes.shape({
-    actions: PropTypes.node,
+  topBarTone: PropTypes.oneOf(['white', 'gray']),
+  topBarLabels: PropTypes.shape({
```
Nit: we often prefer flat props to avoid people passing in new object references. so `topBarCurrentPathLabel` etc. The sum of character needing typing is the same after all.

jonnybel replied:
Yeah, this was one of the things I wasn't so sure myself and was awaiting feedback.

tdeekens commented on packages/application-components/src/components/modal-page/modal-page.js:
```diff
@@ -29,136 +28,74 @@ const parentSelector = () =>
+  children: PropTypes.node.isRequired,
+  showTopBar: PropTypes.bool,
+  showHeader: PropTypes.bool,
   baseZIndex: PropTypes.number,
```
This combincation still makes me worry but I guess it's only me.

jonnybel replied:
It might be useful, but it might also never be used. I guess I can remove it and we add it later if it becomes necessary.

montezume commented on packages/application-components/src/components/modal-page/README.md:
```diff
@@ -14,41 +14,42 @@ Modal pages are controlled components used to render a page using a modal contai
+| `showHeader`     | `boolean`          |          | -                | `true`  | Set this to `false` to not display the header, which contains the title, subtitle and the `headerActions` node.                                                                                                                                                                                             |
+| `headerActions`  | `node`             |          | -                | -       | This node will be rendered in the header section, on the right side of the subtitle, as long as the `showHeader` prop isn't falsy.                                                                                                                                                                          |
+| `topBarLabels`   | `object`           |          |                  | `{}`    | Use this object to pass `currentPath` and `previousPath` fields as labels to the top bar.                                                                                                                                                                                                                   |
+| `topBarTone`     | `string`           |          | [`white`,`gray`] | `white` | Set the colour of the top bar section.                                                                                                                                                                                                                                                                      |
```
I'm not sure I would use `tone`. Why not use `color` or `backgroundColor`?

jonnybel replied:
Fine by me :) I also thought `shade`, since it's only b&w, but yeah `color` is just easier to understand.

tdeekens replied:
Good point. The MC has a history of coming up with any possible name for something like this. `shade` would be something new. We have `tone`, `color` and `theme` and I think just color for these things is fine :)

emmenko commented on packages/application-components/src/components/modal-page/modal-page.js:
```diff
@@ -29,136 +28,74 @@ const parentSelector = () =>
+  subtitle: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
+  children: PropTypes.node.isRequired,
+  showTopBar: PropTypes.bool,
+  showHeader: PropTypes.bool,
```
Have you considered having different components with specific properties instead of one component with many different combinations?
For example, the dialog components have been semantically split into Info,Confirmation,Form components, even though their APIs are very similar as they are built with the same set of internal components. The reason for splitting them is to give them semantic names, which makes it clear when to use what and to keep the respective APIs restrictive in regards to the unique properties of each components.

With that in mind, I would expect something similar to the “modal page” components, for example:
- `InfoModalPage` (the one without actions)
- `FormModalPage` (the one with actions and white header background, without any sub pages)
- `TabularModalPage` (the one with actions and gray header background, with sub pages as tabs)

The names can change, but the idea is what I’m trying to explain.

jonnybel replied:
Yes, I've considered that, even though I first opted for this more "generic" version of this Modal Page, but I've already separated some of the elements into internal components (topBar and Header) with that in mind.

I only left out the Tabular navigation: currently the dev would pass that whole part as the content (children) of the modal, and "turn off" the header.

But if we all agree on this, I can build on top of the current implementation and provide the 3 different, more specific versions of it 🙂

emmenko replied:
Yes, for me it comes down to make the lives of the developers easier, by reducing the cognitive overhead (in this case in terms of correct combinations of props based on the type of usage).

jonnybel replied:
Regarding the `TabularModalPage`: If I understand correctly, if we're gonna integrate the tab-page navigation in this ModalPage, we're gonna need to migrate the Tab components from the MC-FE to app-kit as well.

In fact, I think having a Tab/Tabs component is actually agnostic enough to be a good candidate for UI-kit! What do you think?

montezume replied:
I also think Tab component would fit for UI Kit.

emmenko replied:
To clarify: I didn’t mean to integrate tabs in this component, it’s just a name to suggest the usage of item regards to its layout.
For instance, the FormDialog also does not implement a form, it just indicates that it should be used for form dialogs

jonnybel replied:
Ah, I see, so in this case, the developer would use the "premade" version of `TabularModalPage` which "premade" with the rest of the pieces of the Modal such as TopBar and Header already styled accordingly, and would then nest the Tab navigation and TabContent within the `children` of the modal.

emmenko replied:
Yes, the tabs part is left to the children.

montezume commented on packages/application-components/src/components/modal-page/internals/modal-page.styles.js:
```diff
@@ -0,0 +1,52 @@
+import { css, styled, keyframes } from '@emotion/core';
```
does this work? I thought there was no styled export in `@emotion/core` 🤔 

[ref](https://github.com/emotion-js/emotion/blob/master/packages/core/src/index.js)

the docs say to do `import styled from '@emotion/styled` [docs](https://emotion.sh/docs/styled)

jonnybel replied:
I haven't tested it yet (hence the WIP state). Thanks for pointing this out 😄

montezume commented on packages/application-components/src/components/modal-page/internals/modal-page.styles.js:
```diff
@@ -0,0 +1,52 @@
+  animation: ${fadeIn} 0.2s ease;
+`;
+
+export const ContentWrapper = () => styled.div`
```
why does this need to be a react component instead of just exporting styled.div?

jonnybel replied:
It doesn't 😄

emmenko commented:
What's the status here? Do you need help finishing the implementation?

jonnybel commented:
@emmenko I slowed down a bit over the last two days because of lots of meetings, but I'm trying to wrap this up today, it's all good!

But meanwhile, we also wanted to start deploying the Storybook of app-kit on Netlify so that design can properly test it!

emmenko commented:
> But meanwhile, we also wanted to start deploying the Storybook of app-kit on Netlify so that design can properly test it!

I would like to do that differently, and use the website. We can talk about it and I can show you what you can do with that.

tdeekens commented on packages/application-components/src/components/modal-pages/internals/modal-page-header.js:
```diff
@@ -14,43 +14,50 @@ import filterDataAttributes from '../../../utils/filter-data-attributes';
+      <Text.Subheadline elementType="h4" truncate>
+        {props.title}
+      </Text.Subheadline>
+      {typeof props.subtitle === 'string' ? (
```
What if I pass `14` as a `subtitle`? Maybe using `React.isValidElement` works better?

jonnybel replied:
You are right, `React.isValidElement` is better 👍

jonnybel commented on packages/application-components/src/components/modal-pages/internals/modal-page-header.js:
```diff
@@ -0,0 +1,98 @@
+      css={css`
+        overflow: hidden;
+        & * + * {
+          margin-top: ${customProperties.spacingM} !important;
```
This is basically for the same purpose as using a `<Spacings>` component, but I couldn't use it because I needed to have `overflow: hidden` in order to properly truncate the text of its children.

jonnybel commented:
I've added more VRTs and included examples of the Modal Pages to the new `appkit` website.

emmenko commented:
Thanks, I'll take a look once the deploy preview is ready 🙌

emmenko commented on packages/application-components/src/components/modal-pages/form-modal-page/form-modal-page.story.js:
```diff
@@ -0,0 +1,110 @@
+  TextField,
+  SecondaryButton,
+} from '@commercetools-frontend/ui-kit';
+import { number } from '@storybook/addon-knobs/dist/deprecated';
```
Is this import correct? 🤔

jonnybel replied:
Whoops, VScode automatically did this import and I didn't notice 😄
Good catch!

emmenko commented on packages/application-components/src/components/modal-pages/form-modal-page/form-modal-page.js:
```diff
@@ -0,0 +1,90 @@
+import messages from '../internals/messages';
+
+const useLabel = (label, intl) =>
+  typeof label === 'string' ? label : intl.formatMessage(label);
```
Wait, isn't this wrong? The prop-type says it's either a string or a node, not an intl message 🤔

jonnybel replied:
Hmm, this is used for the props `labelSecondaryButton`, `labelPrimaryButton`, which are either `string` or an `intl` shape. Aren't you confusing it with `subtitle`?

emmenko replied:
You're right, I got confused 🤦‍♂️

We use the same in `dialog-footer.js`, so maybe we can rename it there as well.

emmenko commented on packages/application-components/src/components/modal-pages/form-modal-page/form-modal-page.js:
```diff
@@ -0,0 +1,90 @@
+  labelSecondaryButton: messages.cancel,
+};
+
+FormModalPage.Intl = messages;
```
Careful, you're exposing also messages that should not be used for the button labels, e.g. "back".

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page-header.js:
```diff
@@ -0,0 +1,98 @@
+      justify-content: space-between;
+      margin: 0 ${customProperties.spacingM};
+      padding: ${customProperties.spacingM} 0;
+      border-bottom: 1px solid #afafaf;
```
Why is this color not one of the design tokens?

jonnybel replied:
I coudn't find it 😕 I can have another look, I might have missed it...

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page-top-bar.js:
```diff
@@ -0,0 +1,112 @@
+);
+ModalPageTopBar.displayName = 'ModalPageTopBar';
+ModalPageTopBar.propTypes = {
+  color: PropTypes.oneOf(['white', 'gray']),
```
Can we use other names here? Like `surface` and `neutral`

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page-top-bar.js:
```diff
@@ -0,0 +1,112 @@
+      border-bottom: 1px solid
+        ${props.color === 'gray'
+          ? customProperties.colorWhite
+          : customProperties.colorGray};
```
Those tokens are deprecated, we should use the new ones: `colorSurface` and `colorNeutral`

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page-top-bar.js:
```diff
@@ -0,0 +1,112 @@
+      padding: ${customProperties.spacingS} ${customProperties.spacingM};
+      background-color: ${props.color === 'gray'
+        ? customProperties.colorGray95
+        : customProperties.colorWhite};
```
Same here: `colorNeutral95` and `colorSurface`

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page.styles.js:
```diff
@@ -0,0 +1,53 @@
+import styled from '@emotion/styled';
+import { customProperties } from '@commercetools-frontend/ui-kit';
+
+const slideFromLeft = keyframes`
```
Didn't we say to keep the animation disabled for now?

jonnybel replied:
I've talked with Sven and also showed the animation to Filip here in Berlin and they were quite happy about it and decided it should stay 🙂

emmenko commented on packages/application-components/src/components/modal-pages/form-modal-page/form-modal-page.js:
```diff
@@ -0,0 +1,96 @@
+  labelSecondaryButton: messages.cancel,
+};
+
+FormModalPage.Intl = { confirm: messages.confirm, cancel: messages.cancel };
```
How about we split the messages into 2 files?
- `button-messages.js`
- `messages.js`

Then we can simply `FormModalPage.Intl = buttonMessages;`

jonnybel replied:
I considered it, but preferred this solution to prevent having too many files. But, I have no strong opinion one way or another, so if you prefer to split we can do that.

emmenko replied:
I fear that we forget to update the list here if we add more messages.

Also, we should reuse the same messages that we had in dialogs, to avoid translating them again.

emmenko replied:
Here you go: https://github.com/commercetools/merchant-center-application-kit/pull/642/commits/19a1ed07dada5b647f1881c4084647a66975216d

jonnybel replied:
Thanks 😄

tdeekens commented:
What's the status here. I am entirely lost in the come and go and duration of this. Sorry. Is it ready for review?

emmenko commented:
No, we're working on improving the interactive examples and the VRTs.

emmenko commented:
I opened a PR to extract some of the logic to improve the interactive example setup: #692

I'll take a look also at using react-live, so that people can mess around with the example, if they want to.

montezume commented:
Is it possible to make the website examples not so tiny?

emmenko commented:
> Is it possible to make the website examples not so tiny?

Everything is possible! Keep in mind also that I haven't done a "design review" of the website yet, just focussing on the technical aspects first.

montezume commented on packages/application-components/src/components/dialogs/confirmation-dialog/README.md:
```diff
@@ -31,7 +31,7 @@ Confirmation dialogs are controlled components used to prompt the user to confir
 | Props                           | Type                       | Required | Values            | Default                           | Description                                                                                                                                                                                                                                                                 ...
 | ------------------------------- | -------------------------- | :------: | ----------------- | --------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
-| `title`                         | `string`                   |    ✅    | -                 | -                                 | The title of the Info Dialog                                                                                                                                                                                                                                                 ...
+| `title`                         | `string`                   |    ✅    | -                 | -                                 | The title of the Confirmation Dialog                                                                                                                                                                                                                                         ...
```
😄

jonnybel commented:
It's okay to review now. Even if we add more interactive examples, the core components are the ones which need a proper code review.

emmenko commented:
I added some improvements in the examples. I also noticed that the Tabular component is not implemented correctly, we can talk about it tomorrow.

emmenko commented:
Very nice, the improved examples looks dope!! 🙌

emmenko commented:
The VRT tests also need to be updated

emmenko commented on packages/application-components/src/components/modal-pages/tabular-modal-page/tabular-modal-page.example.js:
```diff
@@ -0,0 +1,174 @@
+      </ExampleWrapper>
+      <Spacings.Inline>
+        <Spacings.Inline alignItems="center">
+          <ToggleInput
```
Nitpick: we should use the small size

emmenko commented:
@jonnybel heads up: I've rebased and I'm fixing some typos in the readmes.

emmenko commented:
VRTs look good now! 🙌

tdeekens commented:
So @emmenko you're not working on this? @jonnybel you're done adding/coordinating things and it can be reviewed?

emmenko commented:
We're working on this yes :)

Abel is taking care of the UX review, I worked on making UX reviews easier (#733 ). I'd like to merge that other PR first and refactor the examples here, so that we can test it properly.

jonnybel commented:
OK, I've rebased and integrated the new "knobs", and thrown away those toggles I had added for using custom controls, and I'm now doing that with the knobs instead.

jonnybel commented:
Design has reviewed and approved.

emmenko commented:
@Shecki @filippobocik can you also review and approve the VRTs? Thanks

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page-header.js:
```diff
@@ -0,0 +1,84 @@
+      justify-content: space-between;
+      margin: 0 ${customProperties.spacingM};
+      padding: ${customProperties.spacingM} 0;
+      border-bottom: 1px solid #afafaf;
```
What about this? We should use a color from uikit tokens

emmenko commented on packages/application-components/src/components/modal-pages/internals/modal-page-top-bar.js:
```diff
@@ -0,0 +1,126 @@
+        margin-left: ${customProperties.spacingS};
+      }
+
+      /* FIXME: these "dirty" styles should be removed when the new Breadcrumbs component is implemented */
```
Is there a task for that already?

jonnybel replied:
There isn't a task as in a "jira ticket", but design said it's planned as an upcoming new feature/component.

emmenko commented on website/src/templates/markdown.js:
```diff
@@ -322,15 +323,17 @@ const components = {
-        </div>
-      </TypographyPage>
-    </MDXProvider>
+    <IntlProvider locale="en">
```
I don't think we need this, it's defined now in the `example-wrapper.js`

jonnybel replied:
Yeah, forgot to remove this after the latest rebase

jonnybel commented:
@emmenko I've addressed the problem of the z-index of the playground InfoDialog, as well as the other remarks.
Let me know if there's something more.

emmenko commented on packages/application-components/src/components/internals/for-docs/example-wrapper/example-wrapper.js:
```diff
@@ -159,6 +160,7 @@ const ExampleWrapper = props => {
                 title="Playground"
                 size="scale"
                 isOpen={isOpen}
+                zIndex={1001}
```
This does not solve the issue.

![image](https://user-images.githubusercontent.com/1110551/59431623-9cd8ec80-8de5-11e9-8f74-bc617f03da25.png)

We need to set this higher, I would put it to `1100`, so like with an offset of `100` (just to be sure).

jonnybel replied:
Oh, I only tested with one opened Modal. 
👍

emmenko commented on packages/application-components/src/components/internals/for-docs/example-wrapper/example-wrapper.js:
```diff
@@ -19,6 +19,7 @@ import { InfoDialog } from '@commercetools-frontend/application-components';
 const PreviewContainer = styled.div`
   position: relative;
   width: 100%;
+  overflow: hidden;
```
What's this for?

jonnybel replied:
It's for the shadows around the portal not to "bleed" out of the PreviewContainer.
![image](https://user-images.githubusercontent.com/2941328/59431908-4e781d80-8de6-11e9-86d9-6f6394799c5a.png)
Design asked about this, but in practice, those shadows won't show on the edge of the pages, only the ones on the left. So I decided to clip them from the PreviewContainer to make it clearer how it will appear on a page.

emmenko replied:
Ah ok, makes sense 👌

jonnybel commented:
I think the VRTs need to be approved again, perhaps because I changed that `border-color` to use one of the design tokens (the "nearest" one) and it's not exactly the same.

emmenko commented:
Ok, should I just approve them again?

jonnybel commented:
Yeah, everything seems to be correct.

emmenko commented:
Done, feel free to finally merge the PR! Thanks a lot for the patience and all the improvements!! 🙌
